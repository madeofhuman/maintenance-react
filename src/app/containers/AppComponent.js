import React from 'react';
import { BrowserRouter, Switch, Route } from 'react-router-dom';
import { Provider } from 'react-redux';
import Homepage from './HomepageContainer';
import NotFound from './404Component';
import store from '../store/store';

const App = () => (
  <Provider store={store}>
    <BrowserRouter>
      <Switch>
        <Route path="/" exact component={Homepage} />
        <Route component={NotFound} />
      </Switch>
    </BrowserRouter>
  </Provider>
);

export default App;
