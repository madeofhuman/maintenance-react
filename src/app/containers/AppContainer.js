import React from 'react';
import { BrowserRouter, Switch, Route } from 'react-router-dom';
import { Provider } from 'react-redux';
import Homepage from './HomepageContainer';
import Dashboard from './DashboardContainer';
import NotFound from './NotFoundContainer';
import ViewRequest from './ViewRequestContainer';
import store from '../store/store';
import PrivateRoute from '../components/PrivateRouteComponent';

export const App = () => (
  <Provider store={store}>
    <BrowserRouter>
      <Switch>
        <Route path="/" exact component={Homepage} />
        <PrivateRoute exact path="/dashboard" component={Dashboard} />
        <PrivateRoute exact path="/admin" component={Dashboard} />
        <PrivateRoute exact path="/view/:requestId" component={ViewRequest} />
        <Route component={NotFound} />
      </Switch>
    </BrowserRouter>
  </Provider>
);

export default App;
