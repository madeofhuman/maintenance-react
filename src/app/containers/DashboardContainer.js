import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Navbar from '../components/NavbarComponent';
import getAllRequests from '../actions/requestActions';
import Loader from '../components/LoaderComponent';
import RequestTable from '../components/RequestTableComponent';
import spannerScrewdriver from '../assets/img/spanner-screwdriver.svg';
import Input from '../components/InputComponent';
import '../assets/css/dashboard.css';

class Dashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  componentDidMount = () => {
    const {
      authenticated, getRequests, history, user,
    } = this.props;
    if (!authenticated && user === undefined) {
      history.push('/');
    }
    if (user.role !== 'admin') {
      history.push('/dashboard');
    }
    if (user.role === 'admin') {
      history.push('/admin');
    }
    getRequests();
  }

  render() {
    const {
      loading, requests, user, message,
    } = this.props;
    return (
      <React.Fragment>
        <Navbar profileIconVisibility="" />
        <div className="body" style={{ backgroundImage: `url(${spannerScrewdriver})` }}>
          <div className="wrapper">
            <div className="left white">
              <span id="username">Welcome, {user.firstName}</span>
            </div>
            <div id="table-sort" className="right white">
              <span id="sort-group">
                <span>Sort by: </span>
                <span>
                  <select className="dropdown" id="request-status-dropdown">
                    <option value="all">All</option>
                    <option value="pending">In-Review</option>
                    <option value="approved">Pending</option>
                    <option value="ongoing">Disapproved</option>
                    <option value="resolved">Resolved</option>
                  </select>
                </span>
                &nbsp;
              </span>
              { user.role !== 'admin' ? (
                <span>
                  <a href="/new">
                    <Input id="new-request-btn" className="button" type="button" value="New Request" />
                  </a>
                </span>)
                : false }
            </div>
            <br />
            <br />
            {requests ? <RequestTable requests={requests} message={message} /> : false}
            <span className="table-nav right hidden">
              <Input type="button" value="<" className="button" />
              <Input type="button" value=">" className="button" />
            </span>
          </div>
          <br />
          <div className="divider" />
        </div>
        {loading ? <Loader /> : false}
      </React.Fragment>
    );
  }
}

Dashboard.defaultProps = {
  user: undefined,
};

Dashboard.propTypes = {
  history: PropTypes.shape({}).isRequired,
  authenticated: PropTypes.bool.isRequired,
  loading: PropTypes.bool.isRequired,
  requests: PropTypes.arrayOf(PropTypes.object).isRequired,
  message: PropTypes.string.isRequired,
  user: PropTypes.shape({}),
  getRequests: PropTypes.func.isRequired,
};

const mapStateToProps = (state, ownProps) => ({
  authenticated: state.auth.authenticated,
  loading: state.common.loading,
  history: ownProps.history,
  requests: state.requests.requests,
  message: state.requests.message,
  user: state.auth.user,
});

const mapDispatchToProps = dispatch => bindActionCreators({
  getRequests: () => (getAllRequests.getAllRequests()),
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Dashboard);
