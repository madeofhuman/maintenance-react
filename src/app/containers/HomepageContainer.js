import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Navbar from '../components/NavbarComponent';
import '../assets/css/home.css';
import wrenchHammer from '../assets/img/wrench-hammer.svg';
import backgroundImage from '../assets/img/gears.svg';
import Modal from '../components/ModalComponent';
import LoginForm from './LoginContainer';
import SignupForm from './SignupContainer';

export class Homepage extends Component {
  state = {
    show: false,
    modalContent: '',
  };

  showModal = (event) => {
    event.preventDefault();
    this.setState({
      show: true,
      modalContent: event.target.name,
    });
  };

  hideModal = () => {
    this.setState({
      show: false,
      modalContent: '',
    });
  };

  componentDidMount = () => {
    const { authenticated, history, role } = this.props;
    if (authenticated) {
      if (role === 'user') {
        history.push('/dashboard');
      }
      if (role === 'admin') {
        history.push('admin');
      }
      if (!role) {
        return true;
      }
    }
  }

  render() {
    const { show, modalContent } = this.state;
    const { history } = this.props;
    return (
      <React.Fragment>
        <Modal show={show} handleClose={this.hideModal}>
          { modalContent === 'sign-in' ? <LoginForm handleClose={this.hideModal} history={history} /> : <SignupForm handleClose={this.hideModal} /> }
        </Modal>
        <Navbar />
        <div className="">
          <div className="body" style={{ backgroundImage: `url(${backgroundImage})` }}>
            <div className="article left">
              <span className="icon">
                <img src={wrenchHammer} alt="icon" />
              </span>
              <span className="text right white">
                <h2>
                  Nobody should have to live with broken things.
                </h2>
                <br />
                <p>
                  We want to help you fix the things you love, and make sure the things that matter
                  most to you, do not fall to pieces, so you can actually enjoy using them.
                  <br /><br />
                  We take great pride in our work and our ability to exceed customer expectation.
                  You, the customer, are our greatest resource and we will treat you accordingly.
                </p>
              </span>
            </div>
            <div id="login-links" className="right">
              <span className="left link-btn">
                <a title="Sign in" name="sign-in" className="white" id="sign-in-link" onClick={this.showModal}>Sign in</a>
              </span>
              <span className="pipe white">
                |
              </span>
              <span className=" link-btn">
                <a title="Sign up" name="sign-up" className="orange" id="sign-up-link" onClick={this.showModal}>Sign up</a>
              </span>
            </div>
          </div>
        </div>
      </React.Fragment>
    );
  }
}

Homepage.defaultProps = {
  role: null,
};

Homepage.propTypes = {
  history: PropTypes.shape({}).isRequired,
  authenticated: PropTypes.bool.isRequired,
  role: PropTypes.string,
};

const mapStateToProps = (state, ownProps) => ({
  authenticated: state.auth.authenticated,
  history: ownProps.history,
  role: state.auth.user !== null ? state.auth.user.role : null,
});

export default connect(mapStateToProps, null)(Homepage);
