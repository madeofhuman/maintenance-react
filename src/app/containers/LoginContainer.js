import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Input from '../components/common/Input';
import authenticate from '../actions/authActions';
import Loader from '../components/common/Loader';

class LoginForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
    };
  }

  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value,
    });
  };

  handleSubmit = (event) => {
    event.preventDefault();
    const {
      email, password,
    } = this.state;
    const { login, history } = this.props;
    const user = {
      email, password,
    };
    return login(user, history, 'login');
  };

  render() {
    const {
      email, password,
    } = this.state;
    const { handleClose, loading } = this.props;
    return (
      <React.Fragment>
        <div id="sign-in-form" className="form">
          <form id="login-form" onSubmit={this.handleSubmit}>
            <span id="sign-up-form-close-btn" className="close-btn grey right" onClick={handleClose}>x</span>
            <p className="output white" />
            <Input
              type="text"
              name="email"
              id="login-email"
              label="Email"
              value={email}
              handleChange={this.handleChange}
              isRequired="required"
            />
            <br /><br />
            <Input
              type="password"
              name="password"
              id="login-password"
              label="Password"
              value={password}
              handleChange={this.handleChange}
              isRequired="required"
            />
            <br /><br />
            <Input
              type="submit"
              value="Log in"
              className="white auth-submit-btn"
              id="signin-btn"
            />
          </form>
        </div>
        { loading ? <Loader /> : '' }
      </React.Fragment>
    );
  }
}

LoginForm.propTypes = {
  handleClose: PropTypes.func.isRequired,
  history: PropTypes.shape({}).isRequired,
  login: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
};

const mapStateToProps = (state, ownProps) => ({
  loading: state.common.loading,
  history: ownProps.history,
});

const mapDispatchToProps = dispatch => bindActionCreators({
  login: (user, history, route) => (authenticate.authenticate(user, history, route)),
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(LoginForm);
