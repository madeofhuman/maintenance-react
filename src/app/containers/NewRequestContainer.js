import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import toastr from 'toastr';
import RequestForm from '../components/RequestForm';
import requestActions from '../actions/requestActions';

/**
 * Manages the state and actions of the New Request Component
 * @class
 * @extends React.Component
 */
export class NewRequest extends Component {
  constructor(props) {
    super(props);
    this.state = {
      type: '',
      item: '',
      model: '',
      detail: '',
    };
  }

  /**
   * Updates the value of the item in the form state with the value of their bound
   * DOM elements when the DOM elements change
   */
  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value,
    });
  }

  /**
   * Validates the user input and calls the createRequest action dispatcher
   * to create the request if the input is valid, or toast an error otherwise.
   */
  handleSubmit = (event) => {
    event.preventDefault();
    const {
      createRequest, getAllRequests,
    } = this.props;
    const {
      type, item, model, detail,
    } = this.state;
    const requestData = {
      type, item, model, detail,
    };
    if (requestData.type.trim() === '') return toastr.error('Type cannot be empty.');
    if (requestData.item.trim() === '') return toastr.error('Item cannot be empty.');
    if (requestData.detail.trim() === '') return toastr.error('Description cannot be empty.');
    if (requestData.model.trim() === '') requestData.model = 'N/A';
    createRequest(requestData);
    getAllRequests();
  }

  /**
   * Renders the New Request Form component on a node in the DOM
   */
  render() {
    const { handleClose } = this.props;
    return (
      <RequestForm
        handleSubmit={this.handleSubmit}
        handleClose={handleClose}
        handleChange={this.handleChange}
      />
    );
  }
}

NewRequest.defaultProps = {
  history: undefined,
};

NewRequest.propTypes = {
  history: PropTypes.shape({}),
  createRequest: PropTypes.func.isRequired,
  getAllRequests: PropTypes.func.isRequired,
  handleClose: PropTypes.func.isRequired,
};

/**
 * Add specified items in the global store as props to the component
 * @param {object} state the global store
 * @param {object} ownProps the component specific props
 */
export const mapStateToProps = (state, ownProps) => ({
  history: ownProps.history,
});

/**
 * Add specified action creators as props to the component
 * @param {function} dispatch - dispatch the specified action.
 */
export const mapDispatchToProps = dispatch => bindActionCreators({
  createRequest: requestData => (requestActions.createRequest(requestData)),
  getAllRequests: () => (requestActions.getAllRequests()),
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(NewRequest);
