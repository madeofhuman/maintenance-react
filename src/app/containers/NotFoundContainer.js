import React from 'react';
import Navbar from '../components/NavbarComponent';
import '../assets/css/404.css';

export const NotFound = () => (
  <React.Fragment>
    <div className="wrapper">
      <Navbar />
      <div className="body" style={{ backgroundImage: "url('../assets/img/gears.svg')" }}>
        <div className="wrapper white">
          <h1 className="center" id="four-o-four-message">
            <span role="img" aria-label="sad face">ðŸ˜¢</span><br />Nothing exists here.
          </h1>
          <em className="center" id="back-btn" onClick={() => (window.history.back())}>&lt;&nbsp;&nbsp;Go Back</em>
        </div>
      </div>
    </div>
  </React.Fragment>
);

export default NotFound;
