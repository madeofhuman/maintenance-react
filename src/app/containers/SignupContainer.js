import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import toastr from 'toastr';
import Input from '../components/common/Input';
import authenticate from '../actions/authActions';
import Loader from '../components/common/Loader';

class SignupForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      firstName: '',
      lastName: '',
      email: '',
      password: '',
      confirmPassword: '',
    };
  }

  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value,
    });
  };

  handleSubmit = (event) => {
    event.preventDefault();
    const {
      firstName, lastName, email, password, confirmPassword,
    } = this.state;
    const { signup } = this.props;
    toastr.options.preventDuplicates = true;
    if (password !== confirmPassword) return toastr.error('passwords don\'t match');
    const user = {
      firstName, lastName, email, password,
    };
    return signup(user, null, 'signup');
  };

  render() {
    const {
      firstName, lastName, email, password, confirmPassword,
    } = this.state;
    const { handleClose, loading } = this.props;
    return (
      <React.Fragment>
        <div id="sign-up-form" className="form">
          <form id="signup-form" onSubmit={this.handleSubmit}>
            <span id="sign-up-form-close-btn" className="close-btn grey right" onClick={handleClose}>x</span>
            <p className="output white" />
            <Input
              type="text"
              name="firstName"
              label="First Name"
              id="first-name"
              value={firstName}
              handleChange={this.handleChange}
              isRequired="required"
            />
            <br /><br />
            <Input
              type="text"
              name="lastName"
              label="Last Name"
              id="last-name"
              value={lastName}
              handleChange={this.handleChange}
              isRequired="required"
            />
            <br /><br />
            <Input
              type="text"
              name="email"
              label="Email"
              id="signup-email"
              value={email}
              handleChange={this.handleChange}
              isRequired="required"
            />
            <br /><br />
            <Input
              type="password"
              name="password"
              label="Password"
              id="signup-password"
              value={password}
              handleChange={this.handleChange}
              isRequired="required"
            />
            <br /><br />
            <Input
              type="password"
              name="confirmPassword"
              label="Confirm Password"
              id="password-confirm"
              value={confirmPassword}
              handleChange={this.handleChange}
              isRequired="required"
            />
            <br /><br />
            <Input
              type="submit"
              value="Sign up"
              className="orange auth-submit-btn"
              id="signup-btn"
            />
          </form>
        </div>
        { loading ? <Loader /> : false }
      </React.Fragment>
    );
  }
}

SignupForm.propTypes = {
  handleClose: PropTypes.func.isRequired,
  signup: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
};

const mapStateToProps = (state, ownProps) => ({
  loading: state.common.loading,
  history: ownProps.history,
});

const mapDispatchToProps = dispatch => bindActionCreators({
  signup: (user, history, route) => (authenticate.authenticate(user, history, route)),
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(SignupForm);
