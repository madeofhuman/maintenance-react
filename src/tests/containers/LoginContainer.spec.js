import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { LoginForm } from '../../app/containers/LoginContainer';

Enzyme.configure({ adapter: new Adapter() });

const props = {
  handleClose: jest.fn(),
  login: jest.fn(),
  clearMessages: jest.fn(),
  history: {
    push: jest.fn(),
  },
  loading: false,
  user: {
    firstName: 'Chukwuka',
    lastName: 'Odina',
    email: 'josephodina@gmail.com',
    role: 'user',
  },
  message: null,
  error: null,
};

describe('Login form component', () => {
  test('renders', () => {
    const wrapper = shallow(<LoginForm {...props} />);
    expect(wrapper.exists()).toBe(true);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
