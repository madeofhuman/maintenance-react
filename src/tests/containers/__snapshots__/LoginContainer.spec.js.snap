// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login form component renders 1`] = `
<Fragment>
  <div
    className="auth-form"
    id="sign-in-form"
  >
    <form
      id="login-form"
      onSubmit={[Function]}
    >
      <span
        className="close-btn grey right"
        id="sign-up-form-close-btn"
      >
        x
      </span>
      <p
        className="output white"
      />
      <Input
        className=""
        handleChange={[Function]}
        id="login-email"
        isRequired="required"
        label="Email"
        name="email"
        type="text"
        value=""
      />
      <br />
      <br />
      <Input
        className=""
        handleChange={[Function]}
        id="login-password"
        isRequired="required"
        label="Password"
        name="password"
        type="password"
        value=""
      />
      <br />
      <br />
      <Input
        className="white auth-submit-btn"
        id="signin-btn"
        isRequired=""
        type="submit"
        value="Log in"
      />
    </form>
  </div>
</Fragment>
`;
